pkg/scanner/nmap.go

```diff
@@ -117,39 +125,81 @@ type nmapContainer struct {
 
 // NMAP implements the scanner interface by making a subprocess call to nmap.
 type NMAP struct {
+	BinPath        string
+	BinArgs        []string
+	Scripts        []string
+	ScriptArgs     []string
 	CommandStrings []string
 	CommandMaker   CommandMaker
 }
 
 // NewNMAP generates a Scanner implementation tha leverages NMAP internally.
 func NewNMAP(binPath string, binArgs []string, scripts []string, scriptArgs []string) *NMAP {
-	strs := make([]string, 0, 6+len(binArgs)+len(scriptArgs))
-	strs = append(
-		strs,
-		binPath,
-		`-sV`,      // Enable version detection of systems.
-		`-oX`, `-`, // Enable XML output mode and send to stdout.
-	)
-	strs = append(strs, binArgs...)
-	if len(scripts) > 0 {
-		strs = append(strs, `--script`, strings.Join(scripts, ","))
+	return &NMAP{
+		BinPath:      binPath,
+		BinArgs:      binArgs,
+		Scripts:      scripts,
+		ScriptArgs:   scriptArgs,
+		CommandMaker: &ExecMaker{},
 	}
-	if len(scriptArgs) > 0 {
-		// Optionally pass in script args if present.
-		strs = append(strs, `--script-args`, strings.Join(scriptArgs, ","))
+}
+
+// WithScripts overrides the active scripts for a scan.
+func (s *NMAP) WithScripts(scripts []string) *NMAP {
+	return &NMAP{
+		BinPath:      s.BinPath,
+		BinArgs:      s.BinArgs,
+		Scripts:      scripts,
+		ScriptArgs:   s.ScriptArgs,
+		CommandMaker: s.CommandMaker,
 	}
+}
+
+// WithScriptArgs overrides the active scripts arguments for a scan.
+func (s *NMAP) WithScriptArgs(args []string) *NMAP {
 	return &NMAP{
-		CommandStrings: strs,
-		CommandMaker:   &ExecMaker{},
+		BinPath:      s.BinPath,
+		BinArgs:      s.BinArgs,
+		Scripts:      s.Scripts,
+		ScriptArgs:   args,
+		CommandMaker: s.CommandMaker,
+	}
+}
+
+// ScanWithScripts overrides the default script and script arguments.
+func (s *NMAP) ScanWithScripts(ctx context.Context, scripts []string, scriptArgs []string, host string) ([]domain.Finding, error) {
+	sc := s
+	if len(scripts) > 0 {
+		sc = sc.WithScripts(scripts)
+	}
+	if len(scriptArgs) > 0 {
+		sc = sc.WithScriptArgs(scriptArgs)
 	}
+	return sc.Scan(ctx, host)
 }
 
 // Scan a host using nmap.
 func (s *NMAP) Scan(ctx context.Context, host string) ([]domain.Finding, error) {
-	cmdStr := make([]string, len(s.CommandStrings)+1)
-	copy(cmdStr, s.CommandStrings)
-	cmdStr[len(cmdStr)-1] = host
-	cmd := s.CommandMaker.MakeCommand(ctx, cmdStr[0], cmdStr[1:]...)
+	strs := make([]string, 0, 6+len(s.BinArgs))
+	strs = append(
+		strs,
+		s.BinPath,
+		`-sV`,      // Enable version detection of systems.
+		`-oX`, `-`, // Enable XML output mode and send to stdout.
+	)
+	strs = append(strs, s.BinArgs...)
+	if len(s.Scripts) > 0 {
+		strs = append(strs, `--script`, strings.Join(s.Scripts, ","))
```

Comment:
Most likely a nit - can we  normalize the path to the script ensure the script(s) really exist on FS level, make sure it is under path where the scripts are supposed to be and not /whetever/path/we/do/not/want ? I understand this comes from internal API, but I still feel not awesome re: running something based on external input.